### PR TITLE
edf: adding a construction option with a given number of pre-picks

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: edf: adding a construction option with a given number of pre-picks
Additional Description:
This is a follow up on #29953, with some insights about an additional scenario that has a similar problem.
The EDF scheduler is used in 2 levels of Zone-Aware LB policies:
1) For choosing the locality to use ([here for healthy hosts](https://github.com/envoyproxy/envoy/blob/5a15b9fc828556e1243c852a24ce3f78d0c332a6/source/common/upstream/upstream_impl.h#L670)). Note that as opposed to the problem in (2), this also impacts equal-weighted localities.
2) For choosing the host within the chosen locality ([here](https://github.com/envoyproxy/envoy/blob/5a15b9fc828556e1243c852a24ce3f78d0c332a6/source/common/upstream/load_balancer_impl.h#L528)).

Prior suggestions ([WRSQ](#14597), #29953) focused on fixing level (2), and required converting the weights to integer-precision before performing the LB policy.
While this may work for (2), assuming one can convert the weights to integers by multiplying by some factor and truncating, for (1) this approach is more challenging as when computing the ["effective locality weight"](https://github.com/envoyproxy/envoy/blob/5a15b9fc828556e1243c852a24ce3f78d0c332a6/source/common/upstream/upstream_impl.cc#L621) its value is dependent on the over-provisioning factor and the ratio of available hosts to all hosts. An additional benefit is that the current approach can also work with slow-start.

Thus, this PR suggests a different mechanism to "perform" some random number of picks when creating the EDF scheduler. The creation process is split into two steps:
1. Estimate a lower-bound on the number of picks each entry will be chosen, and initialize the internal EDF priority-queue accordingly.
2. Perform up to N "pickAndAdd" operations, where N is the number of localities/hosts.

Note that this approach is ~equal to performing some P picks from the scheduler where P >> N (~equal up to double precision computation differences and entries order of entries with the same weight).

After this PR, the next thing is to plumb the PRNG (or a random value) into the locality-scheduler creation process, and fix the call site for (1).

Here's a [short doc](https://docs.google.com/document/d/1i53RYOy8sJUwL6_PSIF4yN39Xc1LvJtZ-18aSBZq0vM/edit?usp=sharing) providing the rationale behind this work.

Risk Level: low - code not used by Envoy's main codebase
Testing: Added tests to validate equivalence with prior approach.
Docs Changes: N/A
Release Notes:N/A - will be updated when the 
Platform Specific Features: N/A

### CODE REVIEW

**Comment by adisuissa** [comment]
cc'ing all the interested stake-holders for the prior fix for comments/suggestions on this new approach
@tonya11en @wbpcode @nezdolik @htuch

**Comment by adisuissa** [comment]
Here's a [short doc](https://docs.google.com/document/d/1i53RYOy8sJUwL6_PSIF4yN39Xc1LvJtZ-18aSBZq0vM/edit?usp=sharing) providing the rationale behind this work.

**Comment by tonya11en** [comment]
Thanks for the doc. This is one of those PRs I'll need to sit down and work through an example before grokking. I've got a block reserved for this tomorrow morning, so give me at least until then.

**Comment by tonya11en** [comment]
Alright, spent some time with it and think it's really clever. It makes sense how you optimize the rotation by skipping to some virtual time in the future and burning through whatever remainder is left. 

I hacked together a "first pick" test similar to the ones done for the WRSQ approach and noticed that the selection probabilities are _not quite_ what I expected. For example, if I have 2 hosts with weights {25, 75} the selection probabilities are not quite right and it doesn't converge on the right answer as I increase the number of iterations in the test:

hosts: {1337, 1338}
weights: {1, 99}
```
[ RUN      ] EdfSchedulerTest.EdfHax
Testing first picks. iters=1000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 21.1%
	expected: 25%
	relative_error: -15.6%
1338:
	observed: 78.9%
	expected: 75%
	relative_error: 5.2%

Testing first picks. iters=10000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 20.36%
	expected: 25%
	relative_error: -18.56%
1338:
	observed: 79.64%
	expected: 75%
	relative_error: 6.18667%

Testing first picks. iters=100000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 20.351%
	expected: 25%
	relative_error: -18.596%
1338:
	observed: 79.649%
	expected: 75%
	relative_error: 6.19867%

Testing first picks. iters=1000000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 20.2892%
	expected: 25%
	relative_error: -18.8432%
1338:
	observed: 79.7108%
	expected: 75%
	relative_error: 6.28107%
```

The unrotated scheduler exhibits the behavior that motivated this change- chooses the same element every time. What I find confusing is that the relative error doesn't change for the scheduler that does the rotations.

I tried this with weights {1, 99} as well and see the same behavior:
```
[ RUN      ] EdfSchedulerTest.EdfHax
Testing first picks. iters=1000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 0.9%
	expected: 1%
	relative_error: -10%
1338:
	observed: 99.1%
	expected: 99%
	relative_error: 0.10101%

Testing first picks. iters=10000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 0.9%
	expected: 1%
	relative_error: -10%
1338:
	observed: 99.1%
	expected: 99%
	relative_error: 0.10101%

Testing first picks. iters=100000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 0.648%
	expected: 1%
	relative_error: -35.2%
1338:
	observed: 99.352%
	expected: 99%
	relative_error: 0.355556%

Testing first picks. iters=1000000
unrotated EDF:
1338: 100%

rotated EDF:
1337:
	observed: 0.6105%
	expected: 1%
	relative_error: -38.95%
1338:
	observed: 99.3895%
	expected: 99%
	relative_error: 0.393434%
```

Any ideas about what's going on here? I can look at this some more tomorrow if not. It's possible I'm doing something weird here, so if you want to take a look at the patch it's here: https://github.com/tonya11en/envoy/commit/bd34ce76e26b5691661a970d907d32f50ed58075

**Comment by adisuissa** [comment]
That's a good observation, and indeed strange.
Let me take a deeper look and add some tests that validate the correct behavior.
/wait-any

**Comment by adisuissa** [comment]
> Any ideas about what's going on here? I can look at this some more tomorrow if not. It's possible I'm doing something weird here, so if you want to take a look at the patch it's here: [tonya11en@bd34ce7](https://github.com/tonya11en/envoy/commit/bd34ce76e26b5691661a970d907d32f50ed58075)

@tonya11en thanks again for providing this test.
The underlying issue was that when there is a weight that is a multiplications of another (say 25 and 75) then the emulated picks can cause some imbalance on the first-pick.
I've solved it by augmenting the weights used, by adding an epsilon to each weight, and added a test based on the one you provided.

File: source/common/upstream/edf_scheduler.h
Comment (tonya11en) [inline source/common/upstream/edf_scheduler.h:94]:
```suggestion
    // weight to avoid cases where weights are multiples of each other.
```

File: source/common/upstream/edf_scheduler.h
Comment (tonya11en) [inline source/common/upstream/edf_scheduler.h:96]: nit: Consider `1e-10`

File: source/common/upstream/edf_scheduler.h
Comment (tonya11en) [inline source/common/upstream/edf_scheduler.h:147]: Should you be passing in `aug_calculate_weight` or `calculate_weight` here? You've already perturbed the weights, so at this point you should use the original weight calculation, right?

File: test/common/upstream/edf_scheduler_test.cc
Comment (tonya11en) [inline test/common/upstream/edf_scheduler_test.cc:232]: I'd throw in an assertion here that verifies `sum(weights) == 100` or change your calculations below to be agnostic to the weight sum.

**Review by tonya11en — APPROVED** [review]
GJ on fixing the first-pick skew! That was a weird one.

LGTM, modulo the minor comments I left.

File: source/common/upstream/edf_scheduler.h
Comment (nezdolik) [reply source/common/upstream/edf_scheduler.h:93]: Can we also have a test for this use case?

**Review by nezdolik — COMMENTED** [review]
Thank you, this is great! Left some nits.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h]: Done.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h]: Good idea, makes it more readable. Done.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h:93]: Added more text.
RE testing, w/o it the first-picks test (the one added after @tonya11en observation) fails.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc]: I think that having a single test that has many random values makes more sense in this case, because at the end of test, it compares the expected and observed values.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc:231]: Done.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h]: You are correct. At the beginning I thought I may need to randomize the order a bit, but as this is not implemented, then the order_offset and the iteration are identical.
Removed `order_offset`.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h]: Done.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc]: Yeah, it does.
Here's an example of an error:
```
test/common/upstream/edf_scheduler_test.cc:37: Failure
The difference between contents1[i].deadline_ and contents2[i].deadline_ is 1.999999966528776e-06, which exceeds 0.000001, where
contents1[i].deadline_ evaluates to 200,
contents2[i].deadline_ evaluates to 199.99999800000003, and
0.000001 evaluates to 9.9999999999999995e-07.
```

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc]: Good idea, done (for all newly introduced tests).

**Comment by adisuissa** [comment]
This is ready for another round of reviews, thanks!

File: test/common/upstream/edf_scheduler_test.cc
Comment (nezdolik) [inline test/common/upstream/edf_scheduler_test.cc:228]:
```suggestion
// Emulates first-pick scenarios by creating a scheduler with the given
```

File: test/common/upstream/edf_scheduler_test.cc
Comment (nezdolik) [inline test/common/upstream/edf_scheduler_test.cc:36]: nit: can we add explanatory comment why do expect deadlines to be equal within absolute error bound `1e-5`?

**Review by nezdolik — COMMENTED** [review]
Few mote nits and ready to ship :)

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc]: Thanks for catching this! Done.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc]: Done

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc:41]: Done.

**Comment by adisuissa** [comment]
Assigning senior maintainers to review.
This is still not plumbed in, but should be reviewed for potential impact in the future.
/assign @htuch @wbpcode

**Review by htuch — COMMENTED** [review]
This is really clever. I want to make sure we can preserve the full reasoning for future readers as algorithms like this need a fast ramp for the uninitiated.

File: source/common/upstream/edf_scheduler.h
Comment (htuch) [reply source/common/upstream/edf_scheduler.h:69]: I also think N is there to show we have a fairly bounded (number of entries) number of picks to do, but the PDF doesn't really talk about that.

File: test/common/upstream/edf_scheduler_test.cc
Comment (htuch) [inline test/common/upstream/edf_scheduler_test.cc:285]: FWIW, given how counter-intuitive some of the quirks we've seen in WRR/EDF, even if just offline, I'd suggest taking  two entries and sweeping incrementally at 0.001 from {100, 0} to {0, 100} or something like that to verify this is solved for good :)

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h:89]: The typical answer is "it depends":
- In [host selection](https://github.com/envoyproxy/envoy/blob/edc366e67472c85c9469f69fe344a612e4b773bd/source/common/upstream/load_balancer_impl.cc#L1086) - it does handle all weights are equal differently.
- In [locality selection](https://github.com/envoyproxy/envoy/blob/edc366e67472c85c9469f69fe344a612e4b773bd/source/common/upstream/upstream_impl.cc#L634) - it doesn't handle this case.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h:96]: I've added a link to the PR comment, but if anyone has a better approach, I'll be happy to revise.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h]: In the initial draft this was called "counts" and later renamed to "picks". I've updated to p'_i.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc:285]: Seems like a test that will have a longish execution time, but I can try to create it.

**Review by adisuissa — COMMENTED** [review]
I'll add the requested test.

File: source/common/upstream/edf_scheduler.h
Comment (htuch) [reply source/common/upstream/edf_scheduler.h:69]: How about taking your Google Doc and putting it in `source/docs` as a `.md`? Looks like GitHub nows supports LaTeX style MArkdown https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/writing-mathematical-expressions.

File: test/common/upstream/edf_scheduler_test.cc
Comment (htuch) [reply test/common/upstream/edf_scheduler_test.cc:285]: Yeah. I think just running it once offline would be fine and having a faster version checked in is fine (e.g. every 0.1). Basically, just want to make sure all the corner cases are gone and there aren't any unknown unknowns.

File: source/common/upstream/edf_scheduler.h
Comment (adisuissa) [reply source/common/upstream/edf_scheduler.h:69]: Sure. I think this should not block this PR.

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc:285]: Unfortunately it can only run ~80 iterations before timing-out (300s).
I'll run it locally but if there's a better idea on testing it, I'd be happy to implement it.

File: source/common/upstream/edf_scheduler.h
Comment (htuch) [reply source/common/upstream/edf_scheduler.h:69]: Sure, I wouldn't block merging, but it would be really nice to have at some point before wrapping up this arc.

File: test/common/upstream/edf_scheduler_test.cc
Comment (htuch) [reply test/common/upstream/edf_scheduler_test.cc:285]: Yeah, maybe just run locally and then check in something with a bunch of points that fits in a reasonable test (just enhancing what is there maybe a bit with extreme points like 0.0001 etc.).

**Review by nezdolik — DISMISSED** [review]

File: test/common/upstream/edf_scheduler_test.cc
Comment (adisuissa) [reply test/common/upstream/edf_scheduler_test.cc:285]: While validating found a strange case (~50 of this out of 100K cases) where the math and floating-point arithmetics did not agree:
floor(A*B) / A was greater than B.
Changed the code a bit to adjust for this. Specifically I've capped the number of picks to ~400M, and add a check for the above corner case, that results in with more iterations for the second part of the algorithm, but keeps the correctness of the EDF scheduler initialization.

PTAL!

File: test/common/upstream/edf_scheduler_test.cc
Comment (htuch) [inline test/common/upstream/edf_scheduler_test.cc:306]: Nit: Exhustive => Exhaustive

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
